couchdb/server/views.py

```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+#
+import copy
+import logging
+
+from couchdb import json
+from couchdb.server.exceptions import QueryServerException, Error
+
+__all__ = ('map_doc',)
+
+log = logging.getLogger(__name__)
+
+
+def map_doc(server, doc):
+    """Applies available map functions to document.
+
+    :command: map_doc
+
+    :param server: Query server instance.
+    :type server: :class:`~couchdb.server.BaseQueryServer`
+
+    :param doc: Document object.
+    :type doc: dict
+
+    :return: List of key-value results for each applied map function.
+
+    :raises:
+        - :exc:`~couchdb.server.exceptions.Error`
+          If any Python exception occurs due mapping.
+    """
+    docid = doc.get('_id')
+    log.debug('Apply map functions to document `%s`:\n%s', docid, doc)
+    orig_doc = copy.deepcopy(doc)
+    map_results = []
+    _append = map_results.append
+    try:
+        for idx, func in enumerate(server.state['functions']):
+            # TODO: https://issues.apache.org/jira/browse/COUCHDB-729
+            # Apply copy.deepcopy for `key` and `value` to fix this issue
+            _append([[key, value] for key, value in func(doc) or []])
```

Comment:
Oh, hmm.. If user write func in normal function not generator.
The right hand side of `func(doc) or []`wil be activate.

e.g.
```python
def func(doc):
    return None
```
works fine.

----
But in case of generator, it need more cost to examine the result... 

I guess the patch looks like this
```diff
diff --git a/couchdb/server/views.py b/couchdb/server/views.py
index 6ce7ac3..7e1d8c7 100644
--- a/couchdb/server/views.py
+++ b/couchdb/server/views.py
@@ -37,7 +37,14 @@ def map_doc(server, doc):
         for idx, func in enumerate(server.state['functions']):
             # TODO: https://issues.apache.org/jira/browse/COUCHDB-729
             # Apply copy.deepcopy for `key` and `value` to fix this issue
-            _append([[key, value] for key, value in func(doc) or []])
+            pairs = []
+            for pair in func(doc) or []:
+                try:
+                    key, val = pair
+                except TypeError:
+                    raise # Error or FatalError?
+                pairs.append([key, val])
+            _append(pairs)
             if doc != orig_doc:
                 log.warning('Document `%s` had been changed by map function'
                             ' `%s`, but was restored to original state',
```